pkg/signature/signature.go

```diff
@@ -0,0 +1,171 @@
+	ErrInvalidEncoding = errors.New("Invalid encoding format - %s")
+
+	// ErrInvalidHash is the error returned when a unsupported hash is supplied.
+	ErrInvalidHash = errors.New("Hash not supported")
+)
+
+type Scheme struct {
+	// Secret represents a list of active secrets used for
+	// a scheme. It is used to implement rolled secrets.
+	// It's order is irrelevant.
+	Secret []string
+
+	Hash     string
+	Encoding string
+}
+
+type Signature struct {
+	Payload json.RawMessage
+
+	// The order of this Schemes is a core part of this API.
+	// We use the index as the version number. That is:
+	// Index 0 = v0, Index 1 = v1
+	Schemes []Scheme
+
+	// This flag allows for backward-compatible implemtation
+	// of this type. You're either generating a simplistic header
+	// or a complex header.
+	Advanced bool
+}
+
+func (s *Signature) ComputeHeaderValue() (string, error) {
+	// Encode Payload
+	tBuf, err := s.encodePayload()
+	if err != nil {
+		return "", err
+	}
+
+	// Generate Simple Signatures
+	if !s.Advanced {
+		sch := s.Schemes[len(s.Schemes)-1]
+		sec := sch.Secret[len(sch.Secret)-1]
+
+		sig, err := s.generateSignature(sch, sec, tBuf)
+		if err != nil {
+			return "", err
+		}
+
+		return sig, nil
+	}
+
+	// Generate Advanced Signatures
+	var hStr strings.Builder
+
+	// Add timestamp.
+	t := fmt.Sprintf("t=%d,", time.Now().Unix())
+	hStr.WriteString(t)
+
+	for k, sch := range s.Schemes {
+		v := fmt.Sprintf("v%d=", k)
+
+		var hSig string
+		for _, sec := range sch.Secret {
+			sig, err := s.generateSignature(sch, sec, tBuf)
+			if err != nil {
+				return "", err
+			}
+
+			hSig = fmt.Sprintf("%s%s,", v, sig)
+			hStr.WriteString(hSig)
+		}
+
+	}
+
+	return hStr.String(), nil
+}
+
+func (s *Signature) generateSignature(sch Scheme, sec string, buf []byte) (string, error) {
+	var sig string
+	var err error
+	switch sch.Encoding {
+	case "hex":
+		if sig, err = s.generateHexSignature(sch.Hash, sec, buf); err != nil {
+			return "", err
+		}
+	case "base64":
+		if sig, err = s.generateBase64Signature(sch.Hash, sec, buf); err != nil {
+			return "", err
+		}
+	default:
+		return "", ErrInvalidEncoding
+	}
+
+	return sig, nil
+}
+
+func (s *Signature) generateHexSignature(hash, secret string, buf []byte) (string, error) {
+	h, err := s.signPayload(hash, secret, buf)
+	if err != nil {
+		return "", err
+	}
+
+	return hex.EncodeToString(h), nil
+}
+
+func (s *Signature) generateBase64Signature(hash, secret string, buf []byte) (string, error) {
+	h, err := s.signPayload(hash, secret, buf)
+	if err != nil {
+		return "", err
+	}
+
+	return base64.StdEncoding.EncodeToString(h), nil
+}
+
+func (s *Signature) signPayload(hash, secret string, buf []byte) ([]byte, error) {
+	fn, err := s.getHashFunction(hash)
+	if err != nil {
+		return nil, err
+	}
+
+	h := hmac.New(fn, []byte(secret))
+	h.Write(buf)
+
+	return h.Sum(nil), nil
+}
+
+func (s *Signature) getHashFunction(algo string) (func() hash.Hash, error) {
```

Comment:
Are these the only hash algorithms we plan on supporting going forward? Because at the moment, we provide support for more than these two. @Subomi

Reply:
@Dotunj The frontend will remove support for others except `SHA256` & `SHA512`